### PR TITLE
Add support for users to upload icons for the project

### PR DESCRIPTION
**Pull request by NingW101**
please test with https://github.com/intel/webinizer-webclient/pull/11

### CODE REVIEW

**Review by lingyuncai — COMMENTED** [review]
Thanks for the efforts, Ning! Below are some general comments from my side.

- For `img` field in `config.json`, I noticed that `server_url` and `project_root` are also stored together. Is it possible to get rid of such info to make the metadata independent from the server deployment and project absolute root path?
  
  For example, for below case with a user uploaded icon, maybe we can just store the image name here. Other info such as `server_url` and `project_root` are all available from the webclient side.
  ```json
  { "img": "http://lingyunc-desk.sh.intel.com:16666/api/projects/%2Fhome%2Flingyuncai%2Fwork%2Fprojects%2Fdev%2Fwebinizer-demo%2Fnative_projects%2Fproject_1/icons/20231023T011916350Z-test" }
  ```
  
  Or for default icon, maybe we can simply use `/assets/icons/default/r.png` and construct the full URL at the webclient side.
  ```json
  { "img": "http://lingyunc-desk.sh.intel.com:16666/assets/icons/default/r.png" }
  ```

-  Maybe it's better to limit the icon image display size to avoid below scenario. This is probably a webclient issue but I just summarized everything here :)
  image

- Have you considered how we handle the uploaded icons in the scenario of publishing/getting packages from the registry?

**Comment by NingW101** [comment]
> Thanks for the efforts, Ning! Below are some general comments from my side.
> 
> * For `img` field in `config.json`, I noticed that `server_url` and `project_root` are also stored together. Is it possible to get rid of such info to make the metadata independent from the server deployment and project absolute root path?
>   For example, for below case with a user uploaded icon, maybe we can just store the image name here. Other info such as `server_url` and `project_root` are all available from the webclient side.
>   ```json
>   { "img": "http://lingyunc-desk.sh.intel.com:16666/api/projects/%2Fhome%2Flingyuncai%2Fwork%2Fprojects%2Fdev%2Fwebinizer-demo%2Fnative_projects%2Fproject_1/icons/20231023T011916350Z-test" }
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   Or for default icon, maybe we can simply use `/assets/icons/default/r.png` and construct the full URL at the webclient side.
>   ```json
>   { "img": "http://lingyunc-desk.sh.intel.com:16666/assets/icons/default/r.png" }
>   ```

Good advice. I'll make modification of hiding the details of server info and project root from the metadata.

> * Maybe it's better to limit the icon image display size to avoid below scenario. This is probably a webclient issue but I just summarized everything here :)
>   image

Of course, I'll add the limitation.
> * Have you considered how we handle the uploaded icons in the scenario of publishing/getting packages from the registry?
In current implementation, users can upload icons when pulling package from the registry and the icons will be stored into the `project_root/.webinizer/icons` folder like `upload project zip` and `clone from github`.
 
As for publishing package, the `img` info is stored and converted to metadata but it is useless because the icon file is not included in the package..
2 possible solutions:
  -  If the `icon` file can be published following the package, we can upload and download the icon file when publishing/pulling package 
  - we do not publish the upload icon info, ignore the img info during converting metadata. When pulling packages from the registry, users can upload and select the icons in current support scenario

**Comment by lingyuncai** [comment]
> > * Have you considered how we handle the uploaded icons in the scenario of publishing/getting packages from the registry?
>
> In current implementation, users can upload icons when pulling package from the registry and the icons will be stored into the `project_root/.webinizer/icons` folder like `upload project zip` and `clone from github`.
> 
> As for publishing package, the `img` info is stored and converted to metadata but it is useless because the icon file is not included in the package.. 2 possible solutions:
> 
> * If the `icon` file can be published following the package, we can upload and download the icon file when publishing/pulling package
> * we do not publish the upload icon info, ignore the img info during converting metadata. When pulling packages from the registry, users can upload and select the icons in current support scenario

I personally prefer the 1st method for user uploaded icons. But some considerations as below.

Currently we ignore the `img` info when converting the metadata.

For user uploaded icon: we can upload the `current` icon stored in `.webinizer/icons` path into the project package, when getting the package from the registry, we can simply set the icon path to `img` field in config.json. This could still work without converting the `img` field into the metadata in `package.json` as only one icon image could be uploaded into the registry.

For project using default icons:
- If the default icons is the corresponding initial letter image - we don't need to upload the icon or convert the `img` field to metadata file `package.json`, and we'll automatically set the initial letter image by default in our current implementation when getting package from the registry.
- If user use a default icon that is not the one of the corresponding initial letter image, this is kind of complex - we needs to either upload the default icon image or convert the `img` field to metadata in `package.json`.

Maybe we can have a discussion on this issue and do it in a following PR when we have a more concrete conclusion, what do you think?

**Comment by NingW101** [comment]
Sounds reasonable to me~ Sure, we can have more discussion about this issue and file another PR to make implementation.

File: src/icons.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+      checkUploadIconType(files.file[0].path);
+      checkIfFileSize(fileSize);
+
+      const targetFilePath = constructIconPath(root, fileName);
+      fs.renameSync(String(files.file[0].path), targetFilePath);
+
+      // update project config
+      const proj = new Project(root);
+      const imgName = path.basename(targetFilePath);
```
Comment (lingyuncai) [inline src/icons.ts:40]: The uploaded icon image is saved without extension names as `20231023T011916350Z-test`.

Maybe use `path.parse(String(files.file[0].originalFilename)).ext` to get the extension and construct the `targetFilePath` together.

**Review by lingyuncai — COMMENTED** [review]
Thanks for the efforts, Ning!

One more comment is that I think it's better to put the demo project's icon image in its own `.webinizer/icons` folder and treat them as the user uploaded icon instead of the default ones.

**Review by lingyuncai — APPROVED** [review]
LGTM, thanks Ning!

**Pull request merged**